Sources/TripKit/managers/TKNotificationManager.swift

```diff
@@ -0,0 +1,142 @@
+//
+//  TKNotificationManager.swift
+//  TripKit
+//
+//  Created by Jules Ian Gilos on 1/16/23.
+//  Copyright © 2023 SkedGo Pty Ltd. All rights reserved.
+//
+
+import Foundation
+
+import UserNotifications
+
+typealias Publisher = ([UNNotificationRequest]) -> Void
+
+public class TKNotificationManager: NSObject {
+  
+  // List down Notification contexts here
+  let subscriptions: [TKNotificationSubscription] = [.init(context: .tripAlerts)]
+  
+  @objc(sharedInstance)
+  public static let shared = TKNotificationManager()
+  
+  /// Clears all the requests for all notification subscriptions
+  public func subscribe(to context: TKNotificationSubscription.Context, updates: @escaping ([UNNotificationRequest]) -> Void) {
+    guard let subscription = getNotificationSubscription(from: context)
+    else {
+      TKLog.warn("Subscribed to a context that is not yet available yet or unrecognized")
+      return
+    }
+    
+    subscription.subscribe(updates)
+  }
+  
+  /// Gets the list of requests that are pending for the provided notification context
+  public func getRequests(from context: TKNotificationSubscription.Context) -> [UNNotificationRequest] {
+    guard let subscription = getNotificationSubscription(from: context)
+    else {
+      return []
+    }
+    
+    return subscription.getRequests()
+  }
+  
+  /// Determines if the provided notification context is subscribed or not
+  public func isSubscribed(to context: TKNotificationSubscription.Context) -> Bool {
+    guard let subscription = getNotificationSubscription(from: context)
+    else {
+      return false
+    }
+    
+    return subscription.isSubscribed()
+  }
+  
+  /// Clears the requests for the provided notification context
+  public func clearRequests(of context: TKNotificationSubscription.Context) {
+    guard let subscription = getNotificationSubscription(from: context)
+    else {
+      return
+    }
+    
+    subscription.clearRequests()
+  }
+  
+  /// Clears all the requests for all notification subscriptions
+  public func clearAllRequests() {
+    subscriptions.forEach { subscription in
+      subscription.clearRequests()
+    }
+  }
+  
+  /// Adds a notification request for the provided context
+  func add(request: UNNotificationRequest, for context: TKNotificationSubscription.Context) {
+    guard let subscription = getNotificationSubscription(from: context)
+    else {
+      TKLog.warn("Adding a request to a context that is not yet available yet or unrecognized")
+      return
+    }
+    
+    subscription.add(request: request)
+  }
+  
+  func getNotificationSubscription(from context: TKNotificationSubscription.Context) -> TKNotificationSubscription? {
+    guard let subscription = subscriptions.first(where: { $0.context == context })
+    else {
+      return nil
+    }
+    
+    return subscription
+  }
+  
+}
+
+public class TKNotificationSubscription {
+  public enum Context {
+    case tripAlerts
+    case none
```

Comment:
Is the `none` case needed, or what's the idea behind that? We could get rid off that if it's not yet used. Could add an `any` case later maybe?

Reply:
I've remove the `none` case for now. Can re-introduce later if needed.

Reply:
ah yes, `any` makes sense more than `none`